### ISSUE TITLE
StringLiteralEquality FN about String literal expression

### ISSUE THREAD

**Issue by Zustin** [checkstyle/checkstyle#11031]
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#StringLiteralEquality
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">    
    <module name="StringLiteralEquality"/>
  </module>
</module>

/var/tmp $ cat Test.java
class Test {
    public void foo() {
        String status = "pending";
        if (status == "do" + "ne") {} // violation
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**

---
Hi, I found a false negative about the rule StringLiteralEquality. At line 4, it is obvious that right value is string literal, but no violation report. I think this is resulted by string literal expression. Thanks for your consideration.

**Comment by Vyom-Yadav** [comment]
@Zustin Thanks for the report, I am on it.

**Comment by strkkk** [comment]
@Vyom-Yadav you have 4 open PRs at the moment, I do not recommend to take new issues if you have a lot of pending PRs.

**Comment by Vyom-Yadav** [comment]
@strkkk Agreed, I will get those merged first and then jump to new issues.

**Comment by strkkk** [comment]
note: this should not be only string concat, text block concat should be considered as well. There is same functionality in EqualsAvoidNullCheck 
https://github.com/checkstyle/checkstyle/blob/7a9913aa64f911e41eae958b757e73a5d2f93205/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheck.java#L385-L406

StringEquality should have same behavior

**Comment by Vyom-Yadav** [comment]
@strkkk What about examples like-
```
String a = "asd";
if(a == "".concat("asd")) { // violation
}
```
I believe that this check could be a little smart and store all String variables (by resolving their scope) and should also be able to identify cases like-
```
String a = "asd";
String b = new String("asd");
if(a == b) { // violation
}
```
Though `b` could be interned somewhere in the program and using `==` would be fine but still, we can show violation for that case.

**Comment by strkkk** [comment]
yes, such enhancement is possible. But definetely outside scope of this issue

**Comment by Vyom-Yadav** [comment]
on it.

**Comment by strkkk** [comment]
fix is merged

**Comment by Vyom-Yadav** [comment]
@strkkk Is it miscellaneous? I think it qualifies as a bug.

**Comment by strkkk** [comment]
my bad, changed
